---
layout: single
published: true
title:  "JAVA 언어 : Couse2. Part3 미니프로젝트 (Feat. 시그니처 백엔드 강의)"
---



## 목차

- [Part 4. 자바 리펙토링 및 미니 프로젝트기](#part-4-자바-리펙토링-및-미니-프로젝트)
  * [Ch03. 실전 API 활용하기 (2)](#ch03-실전-api-활용하기-2)
  * [Ch04. 실전 API 활용하기 (3)](#ch03-실전-api-활용하기-3)

## Part 4. 자바 리펙토링 및 미니 프로젝트

### Ch03. 실전 API 활용하기 (2)


<details>
  <summary>
    close()에 관하여
  </summary>

<div markdown="1">
  <br>

* close() 메서드
  - 메모리에 할당된 인스턴스가 아니라 운영체제의 자원(시스템 자원)을 끌어와서 쓰는 스트림을 닫아줌
  - 자바 외부에 있는 것들 File, Writer Document,  Scanner 등등
  - 그러나 API 중에서 자바 외부가 아니어도 Close가 있는 클래스가 있음(개별확인필요-API문서확인)
  
---

</div>
</details>

<details>
  <summary>
    try 예외처리와 try with resourse 정리
  </summary>

<div markdown="1">
  <br>

* try 예외처리와 try with resourse
  : 수업에서 다루지 않은 예외처리에 대해 상세히 알아보기

  - 우선 예외처리는 오류들을 알려주는 건데, 프로그래밍 언어의 문법적 오류의 경우 __에러(Error)__ 실행중 발생하는 오류 __예외(Exception)__ 이라 한다.
  - throw의 경우 예외를 발생시킬때 사용한다.(사용자에게 주의주거나 관련처리해달라고 부탁)
예시는 다음과같다.

```java
if(x<0){
    throw new Error('x는 음수가 아니어야 합니다.')
}
```

  - catch의 경우 try문에서 예외가 발생하면 호출된다
  - finally의 경우 try문에서 예외가 발생하든 안한든 호출된다.
  - 둘 중 하나는 생략될 수 있다. 둘다 생략은 안된다고한다..
  -  `throws IOException`의 경우 예외가 발생하면 해당 클래스에서 벗어나게 된다.
  - try with resourse 는 `try(Writer writer = new FileWriter(fileName))` 이렇게 괄호안에 스트림에 관한 식하나가 추가되고, 끝나면 닫아준다. try문 안에 있는 예외값을 출력해준다. (finally 쪽이나 stream닫을 때 예욋값을 출력해주려면 다른 추가 코드가 필요하다.
  - try with resourse는 try만 써도 무관하다.

  - try대신 throw new 를 쓰기도하는데,
` if (!response.isSuccessful()) throw new IOException("Request failed: " + response); //응답실패시`
응답실패시 예외를 출력해주게 오류를 발생시켜서 알려준다.(어떤 오류인지 알려주기 위해 사용)

* 예외에 관해서
  : Runtime Excetpion은 자잘한거라서 처리 안해도되는오류
  
  - 예외는 Exception으로 나누며 RuntimeException과 그외로 나뉘는데 그중 IOException은 스트림, 파일 및 디렉터리를 사용하여 정보를 엑세스하는동안 throw된 예외에 대한 기본 클래스이다. > 입출력 작업시 발생되는 오류를 말함
  - throws IOExctiopn은 이오류가 났을시 예외처리
  - 예외처리 되면 밑에는 실행되지 않고 보통 로그로 찍힘


---

</div>
</details>



* 프로젝트 명
  : 크롤링하여 PDF 에 책 목록 만들기

  - https://developers.kakao.com 에 로그인해서  API키 받아오기

 - query 웹 서버에 특정한 정보를 보여달라는 웹 클라이언트 요청




### Ch04. 실전 API 활용하기 (3)





